src/App.tsx

```diff
@@ -1,24 +1,130 @@
-/* eslint-disable max-len */
-/* eslint-disable jsx-a11y/control-has-associated-label */
-import React from 'react';
+import React, {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+
 import { UserWarning } from './UserWarning';
+import { deleteTodo, getTodos, postTodo } from './api/todos';
+import { ErrorInfo } from './components/Errorinfo';
+import { Footer } from './components/Footer';
+import { Header } from './components/Header';
+import { TodoList } from './components/TodoList';
+import { ErrorType } from './types/ErrorType';
+import { FilterType } from './types/FilterType';
+import { NoIdTodo, Todo } from './types/Todo';
+import { filterTodos } from './helper';
 
-const USER_ID = 0;
+const USER_ID = 12017;
 
 export const App: React.FC = () => {
+  const [todosFromServer, setTodosFromServer] = useState<Todo[]>([]);
+  const [selectedFilter, setSelectedFilter] = useState<FilterType>(
+    FilterType.All,
+  );
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMsg, setErrorMsg] = useState<ErrorType | null>(null);
+
+  const showError = (error: ErrorType) => {
+    setErrorMsg(error);
+    setTimeout(() => setErrorMsg(null), 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodosFromServer)
+      .catch(() => showError(ErrorType.TodosNotLoaded));
+  }, []);
+
+  const createNewTodo = (title: string): Promise<void> => {
+    const newTodo: NoIdTodo = {
+      userId: USER_ID,
+      title,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      ...newTodo,
+    });
+
+    return postTodo(newTodo)
+      .then(
+        (response) => setTodosFromServer(
+          currentTodos => [...currentTodos, response],
+        ),
+      )
+      .catch(() => {
+        showError(ErrorType.NotAddable);
+
+        throw new Error(ErrorType.NotAddable);
+      })
+      .finally(() => setTempTodo(null));
+  };
+
+  const removeTodo = (todoToDelete: Todo) => {
+    return deleteTodo(todoToDelete.id)
+      .then(() => setTodosFromServer(
+        currentTodos => currentTodos.filter(todo => todo !== todoToDelete),
+      ))
+      .catch(() => showError(ErrorType.NotDeletable));
+  };
+
+  const clearCompleted = () => {
+    const completedTodos = todosFromServer.filter(todo => todo.completed);
+
+    completedTodos.forEach(completedTodo => {
+      deleteTodo(completedTodo.id)
+        .then(() => setTodosFromServer(
+          currentTodos => currentTodos.filter(todo => todo !== completedTodo),
+        ))
+        .catch(() => showError(ErrorType.NotDeletable));
+    });
+  };
+
+  const filteredTodos = useCallback(
+    filterTodos,
+    [selectedFilter],
+  );
+
+  const todosToView = useMemo(
+    () => filteredTodos(todosFromServer, selectedFilter),
```

Comment:
```suggestion
  const todosToView = useMemo(
    () => filterTodos(todosFromServer, selectedFilter),
```
`filteredTodos` looks redundant, as it does nothing, just wrap `filterTodos` function